bridges/snowbridge/docs/v2.md

```diff
@@ -0,0 +1,310 @@
+
+```solidity
+enum Kind {
+    Index,
+    Address32,
+    Address20,
+    XCMPayload,
+}
+
+struct Beneficiary {
+    Kind kind;
+    bytes data;
+}
+
+enum AssetTransferType {
+	ReserveDeposit, ReserveWithdraw, Teleport
+}
+
+struct Token {
+	AssetTransferType type;
+	address token;
+	uint128 amount;
+}
+
+function send(
+  ParaID destinationChain,
+  Beneficiary calldata beneficiary,
+	Token[] tokens,
+  uint128 reward) external payable;
+```
+
+Message enqueued $m_0$:
+
+```solidity
+send(
+	3022, // KILT Para Id
+	Address32(0x0000....),
+	[
+		(ReserveWithdraw, KLT, 100)
+	],
+	10, // WETH
+	)
+```
+
+```solidity
+send { value: 3 }( // Send 3 Eth for fees and reward
+	3022, // KILT Para Id
+	XCMPayload(
+		DepositAsset (KLT, 100) dest=Bob
+		Transact Bob.hello()
+	),
+	[
+		(ReserveWithdraw, KLT, 100)
+	],
+	1, // 1 ETH of 3 needs to be for the reward, the rest is for fees
+	)
+```
+
+The User Agent/UX layer will need to estimate the fee required to be passed into the `send` method. This may be an issue as we cannot Dry-Run something on Polkadot that has not even been submitted on Ethereum yet. We may need to make RPC API to DryRun and get back the xcm that would be submitted to asset hub.
+
+## Step 2: Relayer relays message to Bridge Hub
+
+On-chain exchange rate is eliminated. Users pay remote delivery costs in ETH, and this amount is sent with the message as WETH. The delivery fee can be claimed by the relayer on BH.
+
+The user agent applies a similar dry-running process as with [Step 2: User agent estimates fees](https://www.notion.so/Step-2-User-agent-estimates-fees-113296aaabef8159bcd0e6dd2e64c3d0?pvs=21).
+
+The message is converted from $m_0$ to $x_0$ during message submission on BH. Dry-running submission will return $x_0$ to the relayer so that it can verify it is profitable.
+
+## Step 3: AH receives $x_0$ from BH
+
+Submitting the message $m_0$ will cause the following XCM, $x_0$, to be built on BH and dispatched to AH.
+
+```
+WithdrawAsset (KLT, 100)
+ReserveAssetDeposited(WETH, 2)
+PayFees (WETH, 1)
+SetAssetClaimer Bob            // derived from beneficiary
+AliasOrigin Ethereum/Alice     // derived from msg.sender
+InitiateAssetsTransfer asset=(KLT, 100) remoteFee=(WETH, 1) dest=KLT
+	DepositAsset (KLT, 100) beneficiary=Bob
+```
+
+```
+WithdrawAsset (KLT, 100)
+ReserveAssetDeposited(WETH, 2)
+PayFees (WETH, 1)
+SetAssetClaimer Ethereum/Alice   // derived from beneficiary
+AliasOrigin Ethereum/Alice       // derived from msg.sender
+InitiateAssetsTransfer asset=(KLT, 100) remoteFee=(WETH, 1) dest=KLT
+		DepositAsset (KLT, 100) beneficiary=Bob
+		Transact Bob.hello()
+```
+
+## Step 4: KILT Receives XCM from AH
+
+The following XCM $x_1$ is received from BH on AH.
+
+```
+*WithdrawAsset (KLT, 100)
+*ReserveAssetDeposited (WETH, 1)
+*PayFees (WETH, 1)
+*SetAssetClaimer Ethereum/Alice
+*AliasOrigin Ethereum/Alice   // origin preserved from AH
+SetAssetClaimer Bob
+DepositAsset (KLT, 100) beneficiary=Bob
```

Comment:
This should be a teleport if we want another message to go to Kilt

Reply:
This is the XCM that KILT will receive from AH. And it assumes reserve-based transfer. If in reality KILT does teleports then `InitiateAssetsTransfer` should take care of doing the right thing here provided we specify the correct reserve type on Asset Hub.